plenum/test/consensus/checkpoint_service/test_checkpoint_service_unit.py

```diff
@@ -0,0 +1,241 @@
+import pytest
+
+from common.exceptions import LogicError
+from plenum.common.constants import DOMAIN_LEDGER_ID
+from plenum.common.messages.internal_messages import Cleanup, StartBackupCatchup, StartMasterCatchup
+from plenum.common.messages.node_messages import Checkpoint, Ordered, PrePrepare, CheckpointState
+from plenum.common.util import updateNamedTuple, getMaxFailures
+from plenum.server.consensus.checkpoint_service import CheckpointService
+from plenum.test.helper import create_pre_prepare_params
+
+
+@pytest.fixture
+def pre_prepare(checkpoint_service):
+    params = create_pre_prepare_params(None,
+                                       ledger_id=DOMAIN_LEDGER_ID,
+                                       view_no=checkpoint_service.view_no,
+                                       pp_seq_no=1)
+    pp = PrePrepare(*params)
+    return pp
+
+
+@pytest.fixture
+def ordered(pre_prepare):
+    ord_args = [
+        pre_prepare.instId,
+        pre_prepare.viewNo,
+        pre_prepare.reqIdr,
+        [],
+        pre_prepare.ppSeqNo,
+        pre_prepare.ppTime,
+        pre_prepare.ledgerId,
+        pre_prepare.stateRootHash,
+        pre_prepare.txnRootHash,
+        pre_prepare.auditTxnRootHash,
+        ["Alpha", "Beta"]
+    ]
+    return Ordered(*ord_args)
+
+
+@pytest.fixture
+def checkpoint(ordered, tconf):
+    start = ordered.ppSeqNo % tconf.CHK_FREQ
+    return Checkpoint(instId=ordered.instId,
+                      viewNo=ordered.viewNo,
+                      seqNoStart=start,
+                      seqNoEnd=start + tconf.CHK_FREQ - 1,
+                      digest='digest')
+
+
+caught_msg = None
+
+
+def catch_msg(msg):
+    global caught_msg
+    caught_msg = msg
+
+
+def test_process_checkpoint_with_incorrect_digest(checkpoint_service, checkpoint, tconf, is_master):
+    key = (checkpoint.seqNoStart, checkpoint.seqNoEnd)
+    sender = "sender"
+    checkpoint_service._checkpoint_state[key] = CheckpointState(1, [],
+                                                                "other_digest", {}, False)
+    assert checkpoint_service.process_checkpoint(checkpoint, sender)
+    if is_master:
+        assert sender not in checkpoint_service._checkpoint_state[key].receivedDigests
+    else:
+        assert sender in checkpoint_service._checkpoint_state[key].receivedDigests
+
+
+def test_process_checkpoint_stash_recvd(checkpoint_service, checkpoint, pre_prepare,
```

Comment:
I think a better name for this test is `test_start_catchup_on_quorum_of_stashed_checkpoints`